// mvc/lib/bootstrap.js
var FrontController = require('./front_controller'),
    Parser = require('./parser'),
    Request = require('./request'),
    Response = require('./response'),
    Router = require('./router');

var Bootstrap = {
    run: function(socket, config){
        socket.id = Math.random();
        socket.parser = new Parser();
        socket.parser.on('request', function(body){
            try{
                var data = JSON.parse(body);
                var request = new Request(body);
                var response = new Response(socket);
                response.setToken(data.token);

                var intent = {
                    socket: socket,
                    config: config,
                    body: data,
                    rawBody: body,
                    url: data.resource,
                    token: data.token,
                    request: request,
                    response: response
                };

                new FrontController(intent).route();
            }catch(ex){
                console.log(ex.stack);
                response.setStatus(403);
                response.write({message: ex});
            }
        });

        socket.on('data', function(buffer){
            this.parser.pushData(buffer);
        });

        socket.on('error', function(err){
            console.log(err);
        });
    }
};
module.exports = Bootstrap;
