// mvc/lib/request.js
var oo = require('./utils/oo');

var Request = function(data){
	this._initRequest(data);
};

Request.prototype = {
	_initRequest: function(data){
		var json = JSON.parse(data);
		this.url = json.resource;
		if(!this.url){
			throw "URL should not be null";
		}
		this.token = json.token;
		this.params = json.params || {};
		this.headers = {};
		this.controllerName = '';
		this.actionName = '';
        this.method = json.method || 'get';

		for(var key in json.headers){
			this.headers[key.toLowerCase()] = json.headers[key];
		}

	},

	addParams: function(params){
		this.params = oo.mix(this.params, params);
	},

	getParam: function(key){
		return this.params[key];
	},

	getUrl: function(){
		return this.url;
	}, 

	getToken: function(){
		return this.token;
	},

	getHeader: function(key){
		return this.headers[key];
	},

    getMethod: function(){
        return this.method.toLowerCase();
    },

	setControllerName: function(value){
		this.controllerName = value;
	},

	getControllerName: function(){
		return this.controllerName;
	},

	setActionName: function(value){
		this.actionName = value;	
	},

	getActionName: function(){
		return this.actionName;	
	}
};

module.exports = Request;
