// mvc/lib/route.js
var stringUtil = require('./utils/string'),
	url = require('url'),
	oo = require('./utils/oo'),
    fs = require('fs');

var Route = function(Options){
	this._initRoute(Options);
}

Route.prototype = {
	_initRoute:function(options){
        this.options = options;
		this._url = stringUtil.trim(options.url, Route.URL_TRIM);

        this.reset(options);

	},

    reset: function(options){
        this._controller = options.controller || 'index';
        this._action = options.action || 'index';
        this._module = options.module || 'default';
        this._params = {};
        this._parts = [];
        this._variables = [];
	    this._partsLength = 0;

        var routePieces = this._url.split(Route.URL_DELIMITER);
        this._length = routePieces.length;


        for(var i = 0, len = routePieces.length; i < len; i++){
            var result = routePieces[i].match(Route.URL_VARIABLE);
            if(result){
                this._parts[i] = null;
                this._variables[i] = result[1];

            }else{
                this._parts[i] = routePieces[i];
	            this._partsLength++;
            }
        }
    },


	match: function(path){
        this.reset(this.options);
		path = stringUtil.trim(path, Route.URL_TRIM);
		var pathPieces = path.split(Route.URL_DELIMITER);
		if(pathPieces.length < this._partsLength){
			return false;
		}
        /*
		//debug info
		console.log(pathPieces);
		console.log(this._variables);
		console.log(this._parts);
		*/
		var len = pathPieces.length > this._parts.length ? pathPieces.length : this._parts.length;
		for(var i = 0; i < len; i++){
			var key = this._variables[i] ? this._variables[i] : null;
			var pathPart = decodeURIComponent(pathPieces[i] ? pathPieces[i] : '');
			if(key === null && pathPart != this._parts[i]){
				return false;
			}
			if(key !== null){
				this._params[key] = pathPart;
				if(key == 'controller' || key == 'action' || key == 'module'){
					if(!!pathPart){
						this['_' + key] = pathPart;
					}

				}
			}
		}

		return true;
	},

	getController: function(){
		return this._controller;
	},

	getAction: function(){
		return this._action;
	},

    getParams: function(){
        return this._params;
    }
}

Route.URL_VARIABLE = /{(\w+)}/;
Route.URL_DELIMITER = /[\/\.]/;
Route.URL_TRIM = '[\/\.]';

module.exports = Route;
